### PR TITLE
Simplify GC prometheus metrics

### PR DESCRIPTION
**Pull request by richvdh**
Rather than hooking into the reactor loop, just add a timed task that runs every 100 ms to do the garbage collection.

Part 1 of a quest to simplify the reactor monkey-patching.

### CODE REVIEW

File: synapse/metrics/_gc.py [hunk 1]
```diff
@@ -81,6 +83,42 @@ def collect(self) -> Iterable[Metric]:
 if not running_on_pypy:
     REGISTRY.register(GCCounts())
 
+    # disable automatic GC, and replace it with a task that runs every 100ms, so that
+    # we can get some metrics about GC activity.
```
Comment (clokep) [inline synapse/metrics/_gc.py:87]: If this is just about gathering metrics, could we use the [`gc.callbacks` mechanism](https://docs.python.org/3/library/gc.html#gc.callbacks) instead and not disable automatic gc?

Comment (richvdh) [reply synapse/metrics/_gc.py:87]: per https://github.com/matrix-org/synapse/issues/9813#issuecomment-1009175679: I don't think so.

Comment (clokep) [reply synapse/metrics/_gc.py:87]: Thanks for answering the same question again. 😢

**Review by squahtx — APPROVED** [review]
lgtm

**Pull request merged**